### PR TITLE
[SPARK-21268][MLLIB] Move center calculations to a distributed map in KMeans

### PR DESCRIPTION
**Pull request by gjgd**
## What changes were proposed in this pull request?

The scal() and creation of newCenter vector is done in the driver, after a collectAsMap operation while it could be done in the distributed RDD.
This PR moves this code before the collectAsMap for more efficiency

## How was this patch tested?

This was tested manually by running the KMeansExample and verifying that the new code ran without error and gave same output as before.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala [hunk 1]
```diff
@@ -292,15 +292,16 @@ class KMeans private (
...
         (sum1, count1 + count2)
-      }.collectAsMap()
+      }
 
       bcCenters.destroy(blocking = false)
 
       // Update the cluster centers and costs
       converged = true
-      totalContribs.foreach { case (j, (sum, count)) =>
+      totalContribs.map { case (j, (sum, count)) =>
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala]: This is OK, but I think I'd keep the distributed operations together above instead.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala [hunk 2]
```diff
@@ -292,15 +292,16 @@ class KMeans private (
       }.reduceByKey { case ((sum1, count1), (sum2, count2)) =>
         axpy(1.0, sum2, sum1)
         (sum1, count1 + count2)
+      }.map { case (j, (sum, count)) =>
+        scal(1.0 / count, sum)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala]: Sorry, I've just thought of a further simplification:

```
.mapValues { case (sum, count) =>
  scal(1.0 / count, sum)
  new VectorWithNorm(sum)
}.collectAsMap()
```

I'll run the tests though

Comment (dardelet) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala]: Good idea ! Updated

**Comment by MLnick** [comment]
Is there any real performance benefit here? Doesn't seem like it unless you have really large `k`?

**Comment by srowen** [comment]
I doubt it makes much difference, yeah. It pushes a little computation off the driver, and computation that happens on each iteration. It seemed like a tiny win that didn't complicate things.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala [hunk 3]
```diff
@@ -292,15 +292,16 @@ class KMeans private (
...
       }.collectAsMap()
 
       bcCenters.destroy(blocking = false)
 
       // Update the cluster centers and costs
       converged = true
-      totalContribs.foreach { case (j, (sum, count)) =>
-        scal(1.0 / count, sum)
-        val newCenter = new VectorWithNorm(sum)
+      totalContribs.foreach { case (j, newCenter) =>
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala]: perhaps the variable name should change from `totalContribs` to `newCenters`?

**Comment by MLnick** [comment]
Fair enough. LGTM with small comment about possibly changing the variable name?

**Comment by dardelet** [comment]
@MLnick Great ! Updated variable name

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**